### PR TITLE
test(deploy): add OCM deploy tool

### LINKED ISSUE #657: Add a new deploytool to test Submariner using OCM

### PR DESCRIPTION
**Pull request by SteveMattar**
Test Submariner using OCM

* Rename the deploytool from olm to bundle
* Create common deploy functions
* Bump olm default version to v0.18.3
* Install `yq` in `shipyard-dapper-base image`
* Add ocm deploy tool
* The new deploy tool is now part of the CI

Closes #657

### CODE REVIEW

File: package/Dockerfile.shipyard-dapper-base [hunk 1]
```diff
@@ -64,6 +65,7 @@ RUN curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/master/i
     mv linux-${ARCH}/helm /go/bin/ && chmod a+x /go/bin/helm && rm -rf linux-${ARCH} && \
     curl -Lo /go/bin/kind "https://github.com/kubernetes-sigs/kind/releases/download/${KIND_VERSION}/kind-linux-${ARCH}" && chmod a+x /go/bin/kind && \
     GOFLAGS="" go get -v github.com/onsi/ginkgo/ginkgo && \
+    GOFLAGS="" go get -v github.com/mikefarah/yq/v4 && \
```
Comment (skitt) [inline package/Dockerfile.shipyard-dapper-base]: Please adjust this to take `YQ_VERSION` into account, and delete line 75.

**Review by skitt — CHANGES_REQUESTED** [review]

**Review by skitt — APPROVED** [review]

File: scripts/shared/deploy.sh [hunk 2]
```diff
@@ -44,10 +44,105 @@ readonly CE_IPSEC_NATTPORT=4500
...
+}
+
+# Create a CatalogSource
+# 1st argument is the catalogsource name
+# 2nd argument is the namespace
+# 3rd argument is index image url
+function create_catalog_source() {
+  local cs=$1
+  local ns=$2
+  # shellcheck disable=SC2034
```
Comment (dfarrell07) [inline scripts/shared/deploy.sh]:
```suggestion
  # shellcheck disable=SC2034 # this variable is used elsewhere
```

File: scripts/shared/deploy.sh [hunk 3]
```diff
@@ -44,10 +44,105 @@ readonly CE_IPSEC_NATTPORT=4500
...
+  local iib=$3
+  echo "[INFO](${cluster}) Create the catalog source ${cs}..."
+
+  kubectl delete catalogsource/operatorhubio-catalog -n "${MARKETPLACE_NAMESPACE}" --wait --ignore-not-found
+  kubectl delete catalogsource/"${cs}" -n "${MARKETPLACE_NAMESPACE}" --wait --ignore-not-found
+
+  # Create the CatalogSource
+  render_template "${RESOURCES_DIR}"/common/catalogSource.yaml | kubectl apply -f -
+
+  # Wait for the catalogSource Readiness
```
Comment (dfarrell07) [inline scripts/shared/deploy.sh]:
```suggestion
  # Wait for the CatalogSource readiness
```

File: scripts/shared/deploy.sh [hunk 4]
```diff
@@ -44,10 +44,105 @@ readonly CE_IPSEC_NATTPORT=4500
...
+  if ! with_retries 60 kubectl get catalogsource -n "${MARKETPLACE_NAMESPACE}" "${cs}" -o jsonpath='{.status.connectionState.lastObservedState}'; then
+    echo "[ERROR](${cluster}) CatalogSource ${cs} is not ready."
+    exit 1
+  fi
+
+  # Debug
+  kubectl -n ${MARKETPLACE_NAMESPACE} get catalogsource --ignore-not-found
+  kubectl -n ${MARKETPLACE_NAMESPACE} get pods --ignore-not-found
+  kubectl -n ${MARKETPLACE_NAMESPACE} get packagemanifests --ignore-not-found | grep "${cs}" || true
```
Comment (dfarrell07) [inline scripts/shared/deploy.sh]: Do we want to keep these in?

File: scripts/shared/deploy.sh [hunk 5]
```diff
@@ -44,10 +44,105 @@ readonly CE_IPSEC_NATTPORT=4500
...
+}
+
+# Create an OperatorGroup
+# 1st argument is the operatorgroup name
+# 2nd argument is the target namespace
+function create_operator_group() {
+  local og=$1
+  local ns=$2
+
+  echo "[INFO](${cluster}) Create the Operator group ${og}..."
```
Comment (dfarrell07) [inline scripts/shared/deploy.sh]:
```suggestion
  echo "[INFO](${cluster}) Create the OperatorGroup ${og}..."
```

File: scripts/shared/deploy.sh [hunk 6]
```diff
@@ -44,10 +44,105 @@ readonly CE_IPSEC_NATTPORT=4500
...
+  # Create the OperatorGroup
+  render_template "${RESOURCES_DIR}"/common/operatorGroup.yaml | kubectl apply -f -
+
+  # Debug
+  kubectl get og -n "${ns}" --ignore-not-found
```
Comment (dfarrell07) [inline scripts/shared/deploy.sh]: Do we want to keep this?

File: scripts/shared/deploy.sh [hunk 7]
```diff
@@ -44,10 +44,105 @@ readonly CE_IPSEC_NATTPORT=4500
...
+  installPlan=$(kubectl get subscriptions.operators.coreos.com "${sub}" -n "${ns}" -o jsonpath='{.status.installPlanRef.name}')
+  if [ -n "${installPlan}" ]; then
+    kubectl patch installplan -n "${ns}" "${installPlan}" -p '{"spec":{"approved":true}}' --type merge
+  fi
+
+  # Debug
+  kubectl get sub -n "${ns}" --ignore-not-found
+  kubectl get installplan -n "${ns}" --ignore-not-found
+  kubectl get csv -n "${ns}" --ignore-not-found
+  kubectl get pods -n "${ns}" --ignore-not-found
```
Comment (dfarrell07) [inline scripts/shared/deploy.sh]: Do we want to keep this?

File: scripts/shared/deploy.sh [hunk 8]
```diff
@@ -44,10 +44,105 @@ readonly CE_IPSEC_NATTPORT=4500
...
+
+  # Delete previous catalogSource and Subscription
+  kubectl delete sub/"${sub}" -n "${ns}" --wait --ignore-not-found
+
+  # Create the Subscription (Approval should be Manual not Automatic in order to pin the bundle version)
+  echo "[INFO](${cluster}) Install the bundle ${bundle} ..."
+  render_template "${RESOURCES_DIR}"/common/subscription.yaml | kubectl apply -f -
+
+  # Manual Approve
+  echo "[INFO](${cluster}) Approve the installPlan..."
```
Comment (dfarrell07) [inline scripts/shared/deploy.sh]:
```suggestion
  echo "[INFO](${cluster}) Approve the InstallPlan..."
```

File: scripts/shared/deploy.sh [hunk 9]
```diff
@@ -44,10 +44,105 @@ readonly CE_IPSEC_NATTPORT=4500
...
+# 3nd argument is the source namespace
+# 4th argument is the bundle name
+function install_bundle() {
+  local sub=$1
+  local cs=$2
+  local ns=$3
+  local bundle=$4
+  local installPlan
+
+  # Delete previous catalogSource and Subscription
```
Comment (dfarrell07) [inline scripts/shared/deploy.sh]:
```suggestion
  # Delete previous CatalogSource and Subscription
```

File: scripts/shared/lib/deploy_ocm [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+readonly OCM_NS="open-cluster-management"
+readonly OCM_AGENT_NS="open-cluster-management-agent"
+readonly HUB="cluster1"
+
+# Variables
+
+### Functions ###
+
+function deploytool_prereqs() {
+  # shellcheck disable=SC2153
```
Comment (dfarrell07) [inline scripts/shared/lib/deploy_ocm]: Can we add explanations for all shellcheck ignores? (example elsewhere)

File: scripts/shared/lib/deploy_ocm [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+  [[ -z "${SUBM_CS}" ]] && (echo "[ERROR] Required environment variables SUBM_CS not loaded"; exit 3)
+  # shellcheck disable=SC2153
+  [[ -z "${SUBM_INDEX_IMG}" ]] && (echo "[ERROR] Required environment variables SUBM_INDEX_IMG not loaded"; exit 3)
+
+  # Create new namespace
+  with_context "${HUB}" "create_namespace ${OCM_NS}"
+  run_subm_clusters "create_namespace ${OCM_AGENT_NS}"
+  run_subm_clusters "create_namespace ${SUBM_NS}"
+
```
Comment (dfarrell07) [inline scripts/shared/lib/deploy_ocm:26]:
```suggestion
```

File: scripts/shared/lib/deploy_ocm [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+}
+
+function create_managed_clusters() {
+  local csr_name
+
+  # Loop the submariner clusters
+  for mc in "${clusters[@]}"; do
+    if [[ ${cluster_subm[$mc]} == "true" ]]; then
+      echo "[INFO](${cluster}) Accept the managed cluster ${mc} and approve its csr..."
+      ### Wait for the managed cluster resource
```
Comment (dfarrell07) [inline scripts/shared/lib/deploy_ocm]:
```suggestion
      # Wait for the managed cluster resource
```

I wouldn't do these triple-comments, for consistency.

File: scripts/shared/lib/deploy_ocm [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+  kubectl apply -k ${submariner_addon_dir}/config/manifests
+}
+
+function create_managed_clusters() {
+  local csr_name
+
+  # Loop the submariner clusters
+  for mc in "${clusters[@]}"; do
+    if [[ ${cluster_subm[$mc]} == "true" ]]; then
+      echo "[INFO](${cluster}) Accept the managed cluster ${mc} and approve its csr..."
```
Comment (dfarrell07) [inline scripts/shared/lib/deploy_ocm]:
```suggestion
      echo "[INFO](${cluster}) Accept the managed cluster ${mc} and approve its Certificate Signing Request (CSR)..."
```

File: scripts/shared/lib/deploy_ocm [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+      ### Wait for the managed cluster resource
+      if ! with_retries 120 kubectl get managedclusters.cluster.open-cluster-management.io "${mc}"; then
+        echo "[ERROR] Unable to find the managed cluster ${mc}."
+        exit 1
+      fi
+      ### Accept the managed cluster
+      kubectl patch managedclusters.cluster.open-cluster-management.io "${mc}" \
+      --type merge --patch '{"spec":{"hubAcceptsClient":true}}'
+      ### Approve the managed cluster csr
```
Comment (dfarrell07) [inline scripts/shared/lib/deploy_ocm]:
```suggestion
      ### Approve the managed cluster CSR
```

File: .github/workflows/scripts.yml [hunk 15]
```diff
@@ -52,15 +52,16 @@ jobs:
...
+          - deploytool: ocm
+            globalnet: globalnet
     steps:
-      - name: Setup kernel for react native, increase watchers
-        if: ${{ matrix.deploytool == 'olm' }}
-        run: |
-          echo -e "fs.inotify.max_user_watches=524288\nfs.inotify.max_user_instances=512" \
-          | sudo tee -a /etc/sysctl.conf && \
-          sudo sysctl -p
+      - name: Setup kernel, increase the inotify settings
```
Comment (dfarrell07) [inline .github/workflows/scripts.yml]:
```suggestion
      - name: Set up kernel, increase the inotify settings
```

File: scripts/shared/lib/deploy_ocm [hunk 16]
```diff
@@ -0,0 +1,167 @@
...
+  echo "[INFO](${cluster}) Create the Managed Cluster Set ..."
+  render_template "${RESOURCES_DIR}"/ocm/managedClusterSet.yaml | kubectl apply -f -
+}
+
+function setup_broker() {
+  echo "[INFO](${cluster}) The broker setup is done by OCM."
+}
+
+function install_subm() {
+  # Loop the submariner clusters and setup the managed cluster addon
```
Comment (dfarrell07) [inline scripts/shared/lib/deploy_ocm]:
```suggestion
  # Loop the submariner clusters and set up the managed cluster addon
```

**Review by dfarrell07 — COMMENTED** [review]
Some suggestions inline. I know some comments are on only-moved code, but it seems to me this is still a good time to improve that code.

**Comment by SteveMattar** [comment]
Thanks for the comments @dfarrell07

**Review by dfarrell07 — APPROVED** [review]
Thanks @SteveMattar!

**Pull request merged**